### ISSUE TITLE
multithread hang while network is not connected

### ISSUE THREAD

**Issue by eaglewangy** [curl/curl#1038]
### I did this
I started two non-block thread
std::async(&Analytics::syncConfiguration, this);
std::async(&Analytics::syncPolicy, this);

I use curl_easy_perform(curl) to perform request in syncConfiguration and syncPolicy,
curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 30) is 30 seconds

I found that if the network is no connected then it will hang the ui(iOS and Android) until the connect timeout.
If the network is connected, everything is ok. or if I set the connect timeout to 1 second, it'll be ok.
How does it happen and how to solve it? thanks
 
### I expected the following
I expect it should not hang the ui(on iOS and android) whatever the connect timeout is.
Because I use asynchronous thread to perform curl request.

### curl/libcurl version
7.50.1

[curl -V output perhaps?]

### operating system
iOS and Android

**Comment by dfandrich** [comment]
What resolver are you using?

**Comment by bagder** [comment]
> or if I set the connect timeout to 1 second, it'll be ok

And if you set the timeout 2 seconds, what happens then? Or 29? Is there a magic cut-off time where it no longer works? What exactly is curl doing at the 30 seconds moment?

**Comment by eaglewangy** [comment]
Sorry for 1 second,  I mean the two threads(or two request) will continue after the timeout.

**Comment by eaglewangy** [comment]
@dfandrich I don't know the resolver. I use the operator's network

**Comment by dfandrich** [comment]
The resolver is a compile-time option. "curl -V", as the bug template suggests, would tell you. If you're using a synchronous resolver with signals disabled, then it will block on the name resolve until its internal timeout hits. But, if you're using the easy API, then it's always blocking by design. If you want non-blocking behaviour, you need to switch to the multi API.

**Comment by eaglewangy** [comment]
resolver:         POSIX threaded
Even I put the easy API in another thread, it will be blocked?

**Comment by eaglewangy** [comment]
And I found another problem: 
I use std::async(&Analytics::syncRemoteConfiguration)
In syncRemoteConfiguration method I use multi API, if the url is interal IP or the url is non-exist(such as: 10.0.18.19/no/such/html), it always blocks the UI thread.

@dfandrich @bagder

**Comment by bagder** [comment]
> resolver: POSIX threaded

That should timeout fine at any time without any signals.

> Even I put the easy API in another thread, it will be blocked?

The easy API is always blocking until its done. But it should time-out as specified. But as long as we can reproduce the problem, we can only offer helps and clues for you to track it down.

> And I found another problem

I don't understand your description of the problem but if it is different it should warrant being a separate issue. But anyway, please elaborate more and describe the problem with many more words!

**Comment by eaglewangy** [comment]
The 'block' means I put the curl request in other thread(not UI thread), it will block the UI thread on Android and iOS.

**Comment by bagder** [comment]
So how can even one thread block another completely independent like that? It doesn't sound like a curl issue to me, it sounds like something else. libcurl has no way to block other threads. On any OS.

**Comment by jzakrzewski** [comment]
Try if you can reproduce the behaviour using
```cpp
auto syncConfigHandle = std::async(std::launch::async, &Analytics::syncConfiguration, this);
auto syncPolicyHandle = std::async(std::launch::async, &Analytics::syncPolicy, this);
```
because I suspect they are called in deferred mode (which is blocking on getting the result). Additionally, be aware that if you do not capture the future returned from the calls, they will block anyway because there will be an implicit future destructor call which waits for the call to finish. Refer to http://en.cppreference.com/w/cpp/thread/async for more details

**Comment by eaglewangy** [comment]
I use 
auto syncConfigHandle = std::async(std::launch::async, &Analytics::syncConfiguration, this);
auto syncPolicyHandle = std::async(std::launch::async, &Analytics::syncPolicy, this);

I can reproduce this problem.
I actually use
 mPolicySyncedResult = std::async(&Analytics::syncPolicy, this);
std::future<bool> mPolicySyncedResult is a class member.
Even only one thread auto syncPolicyHandle = std::async(std::launch::async, &Analytics::syncPolicy, this); it will block the UI thread @jzakrzewski

**Comment by jzakrzewski** [comment]
As @bagder said earlier, if it's really started in a separate thread, there's no way (except some kind of deadlock on mutexes) to lock the calling thread. 
I think you should put some breakpoints and verify that the things are really in separate threads. Additionally you should be able to manually pause the execution when you detect the hang and inspect threads and their callstacks. This can be sometimes quite enlightening (from my experience) ;)

**Comment by eaglewangy** [comment]
I found the problem.
In the two threads, they both perform curl request using a class HttpClient
mHttpClient->performRequest(url, request, responseBody);
In performRequest
size_t HttpClient::performRequest(std::string& url, std::string& body, std::string& responseBdoy)
{
    CURLcode result;
    .......
    result = curl_easy_perform(curl);
    .........
    return responseBdoy.length();
}
It blocks in this code:  result = curl_easy_perform(curl);
If I remove this line, the two threads can go forward without any block

**Comment by bagder** [comment]
So we're back on square one. `curl_easy_perform` is a blocking call that won't return until it is completed.

**Comment by eaglewangy** [comment]
But actually it blocks another thread(UI thread on Android and iOS)
Put curl_easy_perform in a seperate thread, it will block another thread. Only one  curl_easy_perform thread, there is no block. when two curl_easy_perform starts, it blocks the UI thread

**Comment by bagder** [comment]
Yes, but `curl_easy_perform` shouldn't be able to hang any other thread since it has no such super powers.

**Comment by bagder** [comment]
Works for me.

**Comment by eaglewangy** [comment]
@bagder @jzakrzewski @dfandrich 
I find a way to fix it. but the root cause is not clear. Hope get more in-depth answer.

### thread 1
auto syncConfigHandle = std::async(std::launch::async, &Analytics::syncConfiguration, this);
thread ### 2
auto syncPolicyHandle = std::async(std::launch::async, &Analytics::syncPolicy, this);
If I write like above, it will block the third thread(UI thread)

But write to this ---------->
### thread 1
auto syncPolicyHandle = std::async(std::launch::async, &Analytics::syncPolicy, this);
### thread 2
std::thread t(&Analytics::syncRemoteConfiguration, this);
t.detach();
It wont block the third thread(UI thread)

It doesn't relate to curl self, more like multithread problem.
Would you please give some advice?
